src/sqlfluff/core/rules/std/L031.py

```diff
@@ -131,38 +141,70 @@ def _lint_aliases_in_join(
                 continue
 
             alias_identifier_ref = alias_exp_ref.get_child("identifier")
+            yield cls.TableAliasInfo(
+                table_ref, whitespace_ref, alias_exp_ref, alias_identifier_ref
+            )
+
+    def _lint_aliases_in_join(
+        self, base_table, table_expression_segments, column_reference_segments, segment
+    ):
+        """Lint and fix all aliases in joins - except for self-joins."""
+        # A buffer to keep any violations.
+        violation_buff = []
+
+        to_check = list(
+            self._filter_table_expressions(base_table, table_expression_segments)
+        )
+
+        # How many times does each table appear in the FROM clause?
+        table_counts = Counter(ai.table_ref.raw for ai in to_check)
+        table_aliases = defaultdict(set)
+        for ai in to_check:
+            table_aliases[ai.table_ref.raw].add(ai.alias_identifier_ref.raw)
+        for alias_info in to_check:
+            # If the same table appears more than once in the FROM clause with
+            # different alias names, do not consider removing its aliases.
+            # The aliases may have been introduced simply to make each
+            # occurrence of the table independent within the query.
+            if (
```

Comment:
The next few lines are the core of the bug fix. I am not 100% sure what the filter should be here, but I think this is at least a good start. Thoughts?

Reply:
What's the deal with `len(table_aliases[alias_info.table_ref.raw]) > 1`?

Reply:
In the bug report, they were using two separate aliases for the same table. In SQL, two different aliases for the same table allows them to behave "independently" when both are used in the same query, e.g. table `b`, alias `b1` can refer to a different record or set of records than table `b`, alias `b2`. We don't want the lint rule to remove those two aliases, even if they don't appear in the select targets, because then they are forced to reference the same set of records, potentially producing a different result set.

So I'm checking to see if the same table appears with multiple aliases, and that prevents removing the aliases.

Does this answer your question? I know this is kind of a subtle and unusual SQL thing, so I'm happy to continue the conversation as needed.

Reply:
Ah I see what's going on here now. I'm familiar with this practice in SQL but the defaultdict(set) part bypassed me initially. Thanks!